model/invoice.rb

```diff
@@ -15,58 +15,69 @@ def name
   end
 
   def charge
+    reload # Reload to get the latest status to avoid double charging
     unless (Stripe.api_key = Config.stripe_secret_key)
-      puts "Billing is not enabled. Set STRIPE_SECRET_KEY to enable billing."
-      return
+      Clog.emit("Billing is not enabled. Set STRIPE_SECRET_KEY to enable billing.")
+      return true
     end
 
     if status != "unpaid"
-      puts "Invoice[#{ubid}] already charged: #{status}"
-      return
+      Clog.emit("Invoice already charged.") { {invoice_already_charged: {ubid: ubid, status: status}} }
+      return true
     end
 
-    if content["cost"] < Config.minimum_invoice_charge_threshold
+    amount = content["cost"].to_f.round(2)
+    if amount < Config.minimum_invoice_charge_threshold
       update(status: "below_minimum_threshold")
-      puts "Invoice[#{ubid}] cost is less than minimum charge cost: $#{content["cost"]}"
-      return
+      Clog.emit("Invoice cost is less than minimum charge cost.") { {invoice_below_threshold: {ubid: ubid, cost: amount}} }
+      return true
     end
 
-    unless (billing_info = BillingInfo[content.dig("billing_info", "id")])
-      puts "Invoice[#{ubid}] doesn't have billing info"
-      return
+    unless (billing_info = BillingInfo[content.dig("billing_info", "id")]) && billing_info.payment_methods.count > 0
+      Clog.emit("Invoice doesn't have billing info.") { {invoice_no_billing: {ubid: ubid}} }
+      return false
     end
 
-    payment_methods = billing_info.payment_methods_dataset.order(:order).all
-
-    payment_methods.each do |payment_method|
-      amount = content["cost"].to_f.round(2)
+    errors = []
+    success = false
+    billing_info.payment_methods_dataset.order(:order).each do |pm|
       payment_intent = Stripe::PaymentIntent.create({
         amount: (amount * 100).to_i, # 100 cents to charge $1.00
         currency: "usd",
         confirm: true,
         off_session: true,
         customer: billing_info.stripe_id,
-        payment_method: payment_method.stripe_id
+        payment_method: pm.stripe_id
       })
 
       if payment_intent.status == "succeeded"
-        puts "Invoice[#{ubid}] charged with PaymentMethod[#{payment_method.ubid}] for $#{amount}"
+        Clog.emit("Invoice charged.") { {invoice_charged: {ubid: ubid, payment_method: pm.ubid, cost: amount}} }
         self.status = "paid"
         content.merge!({
           "payment_method" => {
-            "id" => payment_method.id,
-            "stripe_id" => payment_method.stripe_id
+            "id" => pm.id,
+            "stripe_id" => pm.stripe_id
           },
           "payment_intent" => payment_intent.id
         })
         save(columns: [:status, :content])
-        return payment_intent.id
+        success = true
+        break
```

Comment:
Instead of break, you can return `true` here. In that way, you wouldn't need to keep a flag for `success`

Reply:
Thanks. Makes sense